reports-backend/src/main/java/org/jboss/da/products/impl/AggregatedProductProvider.java

```diff
@@ -0,0 +1,230 @@
+package org.jboss.da.products.impl;
+
+import org.jboss.da.listings.model.ProductSupportStatus;
+import org.jboss.da.model.rest.GA;
+import org.jboss.da.products.api.Artifact;
+import org.jboss.da.products.api.Product;
+import org.jboss.da.products.api.ProductArtifacts;
+import org.jboss.da.products.api.ProductProvider;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Any;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+
+/**
+ *
+ * @author Honza Brázdil <janinko.g@gmail.com>
+ */
+@ApplicationScoped
+public class AggregatedProductProvider implements ProductProvider {
+
+    public static Set<ProductArtifacts> filterArtifacts(Set<ProductArtifacts> artifacts,
+            Predicate<? super Artifact> predicate) {
+        return artifacts.stream()
+                .map(a -> new ProductArtifacts(a.getProduct(),
+                                a.getArtifacts().stream()
+                                .filter(predicate)
+                                .collect(Collectors.toSet())))
+                .filter(a -> !a.getArtifacts().isEmpty())
+                .collect(Collectors.toSet());
+    }
+
+    public static CompletableFuture<Set<ProductArtifacts>> filterProducts(
+            CompletableFuture<Set<ProductArtifacts>> artifacts,
+            Predicate<? super Product> predicate) {
+        return artifacts.thenApply(as -> as.stream()
+                .filter(a -> predicate.test(a.getProduct()))
+                .collect(Collectors.toSet()));
+    }
+
+    // TODO: filter unknown products, so that when there is artifact both in unknown product and in
+    // known product, it's removed from the unknown
```

Comment:
when we have an information that product was build in a specific product, we don't get any new information knowing it was build in unknown product (as it may be the specific one) and probably every time there is some specific, there is also unknown